File: src/test/java/io/protobase/f7/formulas/math/COTTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.DivException;
import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class COTTest extends TestFormula {
  @Test
  public void testApply() {
    assertThat(COT.SELF.apply(null, 10.0)).isEqualTo(1 / Math.tan(10.0));
    assertThat(COT.SELF.apply(null, 128731.2)).isEqualTo(1 / Math.tan(128731.2));
    assertThat(COT.SELF.apply(null, 11.11)).isEqualTo(1 / Math.tan(11.11));
    assertThat(COT.SELF.apply(null, 0.0)).isEqualTo(new DivException());
    assertThat(COT.SELF.apply(null, 88281.0)).isEqualTo(1 / Math.tan(88281));
    assertThat(COT.SELF.apply(null, 2.0)).isEqualTo(1 / Math.tan(2.0));
    assertThat(COT.SELF.apply(null, 4.0)).isEqualTo(1 / Math.tan(4.0));
    assertThat(COT.SELF.apply(null, -4.0)).isEqualTo(1 / Math.tan(-4.0));
    assertThat(COT.SELF.apply(null, -10124.0)).isEqualTo(1 / Math.tan(-10124.0));
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(COT.SELF.apply(null, "10.0")).isEqualTo(1 / Math.tan(10.0));
  }

  @Test
  public void testApply_grid() {
    assertThat(COT.SELF.apply(null,
        Grid.builder().add(0, 0, -4.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(1 / Math.tan(-4.0));
  }

  @Test
  public void test_lookup() {
    COT F = new COT(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(10.0);
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(1 / Math.tan(10.0));
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(COT.SELF.apply(null, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(COT.SELF.apply(null, "A", "Too many")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/math/COTTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { COT } from "../../../../main/js/formulas/math/COT";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { Converters } from "../../../../main/js/utils/Converters";
import { CommonModels } from "../../testutils/CommonModels";

describe("COT", function () {
  it("should work with valid numbers", function () {
    assert.deepEqual(COT.SELF.run(null, 10.0), 1 / Math.tan(10.0));
    assert.deepEqual(COT.SELF.run(null, 128731.2), 1 / Math.tan(128731.2));
    assert.deepEqual(COT.SELF.run(null, 11.11), 1 / Math.tan(11.11));
    assert.deepEqual(
      Converters.castAsF7Exception(COT.SELF.run(null, 0.0)).name,
      F7ExceptionName.DIV
    );
    assert.deepEqual(COT.SELF.run(null, 88281.0), 1 / Math.tan(88281));
    assert.deepEqual(COT.SELF.run(null, 2.0), 1 / Math.tan(2.0));
    assert.deepEqual(COT.SELF.run(null, 4.0), 1 / Math.tan(4.0));
    assert.deepEqual(COT.SELF.run(null, -4.0), 1 / Math.tan(-4.0));
    assert.deepEqual(COT.SELF.run(null, -10124.0), 1 / Math.tan(-10124.0));
  });

  it("should work with strings", function () {
    assert.equal(COT.SELF.run(null, "1"), 1 / Math.tan(1));
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(COT.SELF.run(null, new ValueException()), new ValueException());
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new COT(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(0.99);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), 1 / Math.tan(0.99));
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 0.99).add(0, 1, "A").build();
    assert.deepEqual(COT.SELF.run(null, one), 1 / Math.tan(0.99));
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((COT.SELF.run(null) as NAException).name, F7ExceptionName.NA);
    assert.deepEqual((COT.SELF.run(null, "A", "B") as NAException).name, F7ExceptionName.NA);
  });
});
